File: scala-snippets/src/main/java/de/richargh/billiondollar/Bar.java
package de.richargh.billiondollar;

import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Bar {

    private final Map<DrinkId, Drink> allDrinks;

    public Bar(Drink... drinks){
        this.allDrinks = Stream.of(drinks).collect(Collectors.toMap(Drink::id, item -> item));
    }

    public Drink findById(DrinkId id){
        return allDrinks.get(id);
    }

    public void place(Drink drink){
        allDrinks.put(drink.id(), drink);
    }

}

File: fsharp-snippets/Main/Bar.fs
module Richargh.BillionDollar.Main.Bar

open System

type CreatorId = {
    RawValue: string
}

[<NoEquality; NoComparison>]
type Creator = {
    Id: CreatorId
    Name: string
}

type DrinkId = {
    RawValue: string
}

[<NoEquality; NoComparison>]
type Drink = {
    Id: DrinkId
    Name: string
    CreatorId: CreatorId option 
}

type IBar =
    abstract All: unit -> Drink list
    abstract FindById: DrinkId -> Drink option
    abstract Put: Drink -> unit

type InMemoryBar([<ParamArray>] drinks: Drink[]) =
    
    let mutable allDrinks: Map<DrinkId, Drink> =
            drinks
            |> Array.map (fun drink -> (drink.Id, drink))
            |> Map.ofArray 

    interface IBar with
        member this.All() = 
            allDrinks
            |> Map.toList
            |> List.map(fun (_, drink) -> drink)

        member this.FindById id = 
            allDrinks.TryFind id

        member this.Put (drink: Drink) = 
            allDrinks <- allDrinks.Add(drink.Id, drink)
            
let barOf ([<ParamArray>] drinks: Drink[]): IBar = upcast InMemoryBar(drinks)